WIP: RS/AS split and PRM (#338)
modelcontextprotocol/modelcontextprotocol: docs/specification/draft/basic/authorization.mdx

implements a selected subset of their features to ensure security and interoperability
while maintaining simplicity:

- OAuth 2.1 IETF DRAFT ([draft-ietf-oauth-v2-1-12](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-12))
- OAuth 2.0 Authorization Server Metadata
  ([RFC8414](https://datatracker.ietf.org/doc/html/rfc8414))
- OAuth 2.0 Dynamic Client Registration Protocol
  ([RFC7591](https://datatracker.ietf.org/doc/html/rfc7591))

> - OAuth 2.0 Protected Resource Metadata IETF DRAFT ([draft-ietf-oauth-resource-metadata-13](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-resource-metadata-13))

Reviewer: This spec has been assigned RFC 9728 but hasn't yet been published. Plan on updating this reference in the next week or so: https://www.rfc-editor.org/auth48/rfc9728

## 2. Authorization Flow

### 2.1 Overview

1. MCP authorization servers **MUST** implement OAuth 2.1 with appropriate security
   measures for both confidential and public clients.

1. MCP authorization servers and MCP clients **SHOULD** support the OAuth 2.0 Dynamic Client Registration
   Protocol ([RFC7591](https://datatracker.ietf.org/doc/html/rfc7591)).

1. MCP servers **MUST** implement [OAuth 2.0 Protected Resource Metadata](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-resource-metadata-13).

>    MCP clients **MUST** use OAuth 2.0 Protected Resource Metadata for authorization discovery.

Reviewer 2: Missing a word?

```suggestion
   MCP clients **MUST** use OAuth 2.0 Protected Resource Metadata for authorization server discovery.
```

>    - For instance, an agent calls a secure MCP tool to check inventory at a specific

Reviewer 2: I am not sure we want to introduce the term "agent" in this section. Should it be "MCP client" for consistency?

Author: This is probably worth a different PR since it wasn't a changed line.

     store. No need to impersonate the end user.

### 2.2 Roles
A protected MCP server acts as a [OAuth 2.1 resource server](https://www.ietf.org/archive/id/draft-ietf-oauth-v2-1-12.html#name-roles),
capable of accepting and responding to protected resource requests using access tokens.

> An MCP client acts a [OAuth 2.1 client](https://www.ietf.org/archive/id/draft-ietf-oauth-v2-1-12.html#name-roles),

Reviewer 2: ```suggestion
An MCP client acts as an [OAuth 2.1 client](https://www.ietf.org/archive/id/draft-ietf-oauth-v2-1-12.html#name-roles),
```

guidance on implementation details.

MCP servers **MUST** use the HTTP header `WWW-Authenticate` when returning a _401 Unauthorized_ to indicate the location of the resource server metadata URL
as described in [OAuth 2.0 Protected Resource Metadata](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-resource-metadata-13).

MCP clients **MUST** be able to parse `WWW-Authenticate` headers and respond appropriately to `HTTP 401 Unauthorized` responses from the MCP server.

#### 2.3.2 Server Metadata Discovery

MCP clients **MUST** follow the OAuth 2.0 Authorization Server Metadata protocol defined
in [RFC8414](https://datatracker.ietf.org/doc/html/rfc8414) to obtain the information
required to interact with the authorization server.

#### 2.3.3 MCP specific headers for discovery

MCP clients **SHOULD** include the `MCP-Protocol-Version: <protocol-version>` HTTP header during
Server Metadata Discovery to allow the MCP server to respond based on the MCP protocol
version.

MCP servers **SHOULD** use the `MCP-Protocol-Version` header to determine compatibility with the MCP client.

> For example: `MCP-Protocol-Version: 2024-11-05`

Reviewer 2: Nit: should the example reference the latest spec revision?

#### 2.3.4 Sequence Diagram
The following diagram outlines an example flow:

```mermaid
sequenceDiagram
    participant C as Client
    participant M as MCP Server (Resource Server)
    participant A as Authorization Server

    C->>M: MCP Request without token
    M-->>C: HTTP 401 Unauthorized with WWW-Authenticate header
    Note over C: Extract resource_metadata<br />from WWW-Authenticate

    C->>M: GET /.well-known/oauth-protected-resource
    M-->>C: Resource metadata with authorization server URL
    Note over C: Validate RS Metadata,<br />build AS metadata URL

    C->>A: GET /.well-known/oauth-authorization-server
    A-->>C: Authorization server metadata

>     Note over C,A: OAuth 2.1 authorization flow happens here

Reviewer: Authentication happens outside the scope of the OAuth flow, so "authentication and authorization" is not accurate. You could explain that the user authentication happens during this stage, but it only happens at the authorization server, not between the client and authorization server. I don't think it's relevant enough to mention in this diagram though, as this should focus on client-AS interactions.

    C->>A: Token Request
    A-->>C: Access Token

    C->>M: MCP Request with Access Token

>     M-->>C: Protected Resource Response

Reviewer: There is a mismatch between "MCP Request" followed by "Protected Resource Response". These should probably be made consistent, either "MCP Request"/"MCP Response" or "Protected Resource Request"/"Protected Resource Response"

    Note over C,M: MCP communication continues with valid token
```

### 2.4 Dynamic Client Registration

> 

Reviewer 3: Nit: update MCP servers to MCP authorization servers in this section (lines 147, 150 and 158).

Author: Good catch, thank you

> 

Reviewer: The two mentions of "client secret" can be changed to "client credentials" to allow for the use of public/private key credentials instead.

sequenceDiagram
    participant B as User-Agent (Browser)
    participant C as Client
    participant M as MCP Server (Resource Server)
    participant A as Authorization Server

    C->>M: MCP Request without token
    M->>C: HTTP 401 Unauthorized with WWW-Authenticate header

>     Note over C: Extract resource_metadata from WWW-Authenticate

Reviewer: ```suggestion
    Note over C: Extract resource_metadata URL from WWW-Authenticate
```

>     alt WWW-Authenticate header not present

Reviewer: ```suggestion
```

The `WWW-Authenticate` response is required as described above.

        C->>M: GET /.well-known/oauth-protected-resource
        M->>C: Resource metadata with authorization_server URL

>     end

Reviewer: ```suggestion
```

The `WWW-Authenticate` response is required as described above.

    C->>A: GET /.well-known/oauth-authorization-server
    A->>C: Authorization server metadata

    alt Dynamic Client Registration
        C->>A: POST /register
        A->>C: Client Credentials
    end

    Note over C: Generate PKCE Parameters
    C->>B: Open browser with authorization URL + code_challenge
    B->>A: Authorization Request
    Note over A: User authorizes
    A->>B: Redirect to callback with authorization code
    B->>C: Authorization code callback
    C->>A: Token Request + code_verifier
    A->>C: Access Token (+ Refresh Token)

>     C->>M: MCP Request with Access Token

Reviewer: Same note about "MCP Request" vs "Protected Resource Response" as above.

error handling requirements. Invalid or expired tokens **MUST** receive a HTTP 401
response.

> MCP client **MUST NOT** send tokens other than issued by the MCP authorization server.

Reviewer 2: ```suggestion
MCP clients **MUST NOT** send tokens to the MCP server other than ones issued by the MCP server's authorization server.
```

> MCP client **MUST NOT** send tokens other than issued by the MCP authorization server.

Author: We are using MCP authorization server throughout the document to refer to the AS responsible for the MCP server. Happy to find a better phrasing, but we need to be consistent across the document.

> MCP servers **MUST** only issue tokens that are valid for use with their own resources.

Author: Yes that's a typo. Thank you!

> MCP servers **MUST** only issue tokens that are valid for use with their own resources.

Reviewer: ```suggestion
MCP authorization servers **MUST** only issue tokens that are valid for use with their own resource servers.
```

### 2.9 Implementation Requirements

1. Implementations **MUST** follow OAuth 2.1 security best practices

> 1. PKCE is **REQUIRED** for all clients

Reviewer 2: ```suggestion
1. PKCE is **REQUIRED** for all MCP clients
```

Author: whoopsy. thanks!

> 1. PKCE is **REQUIRED** for all MCP clients

Reviewer: ```suggestion
1. PKCE is **REQUIRED** for all MCP clients and authorization servers
```

> 1. MCP clients or MCP servers that also act as an AS

Reviewer 2: I don't think we expect clients to be the AS?

```suggestion
1. MCP servers that also act as an AS:
```

>     1. **SHOULD** implement token rotation for enhanced security

Reviewer: Why is this a requirement unique to a tightly coupled AS/RS?

Reviewer 2: @aaronpk if you're using a third-party AS, you would have no control over token rotation policies, no? That is, until you actually have to refresh your own cached tokens.

Reviewer: ok we'll come back to this in a follow-up PR. I'd rather take out this recommendation entirely since the recommendation is really to follow OAuth 2.1's requirements on refresh token rotation.

>     1. Token lifetimes **SHOULD** be limited based on security requirements

Reviewer 2: Consistency in formatting/verbiage:

```suggestion
    1. **SHOULD** restrict token lifetimes based on security requirements
```
